source/io/Stream.cpp

```diff
@@ -13,132 +13,172 @@
-    impl->stream = nullptr;
-    aws_mem_release(stream->allocator, stream);
-}
+                if (!m_underlying_stream)
+                {
+                    m_good = false;
+                    return;
+                }
 
-static struct aws_input_stream_vtable s_aws_input_stream_cpp_vtable = {s_aws_input_stream_cpp_seek,
-                                                                       s_aws_input_stream_cpp_read,
-                                                                       s_aws_input_stream_cpp_get_status,
-                                                                       s_aws_input_stream_cpp_get_length,
-                                                                       s_aws_input_stream_cpp_destroy};
+                m_good = true;
+                m_underlying_stream->impl = this;
+                m_underlying_stream->allocator = m_allocator;
+                m_underlying_stream->vtable = &s_vtable;
+            }
 
-namespace Aws
-{
-    namespace Crt
-    {
-        namespace Io
-        {
-            struct aws_input_stream *AwsInputStreamNewCpp(
-                const std::shared_ptr<Aws::Crt::Io::IStream> &stream,
+            StdIOStreamInputStream::StdIOStreamInputStream(
+                std::shared_ptr<Aws::Crt::Io::IStream> stream,
                 Aws::Crt::Allocator *allocator) noexcept
+                : InputStream(allocator), m_stream(std::move(stream))
             {
-                struct aws_input_stream *input_stream = NULL;
-                struct aws_input_stream_cpp_impl *impl = NULL;
-
-                aws_mem_acquire_many(
-                    allocator,
-                    2,
-                    &input_stream,
-                    sizeof(struct aws_input_stream),
-                    &impl,
-                    sizeof(struct aws_input_stream_cpp_impl));
-
-                if (!input_stream)
+            }
+
+            bool StdIOStreamInputStream::Read(ByteBuf &buffer) noexcept
+            {
+                m_stream->read(reinterpret_cast<char *>(buffer.buffer + buffer.len), buffer.capacity - buffer.len);
+                buffer.len += static_cast<size_t>(m_stream->gcount());
+
+                return true;
+            }
+
+            StreamStatus StdIOStreamInputStream::GetStatus() const noexcept
+            {
+                StreamStatus status;
+                status.is_end_of_stream = m_stream->eof();
+                status.is_valid = m_stream->good();
+
+                return status;
+            }
+
+            int64_t StdIOStreamInputStream::GetLength() const noexcept
+            {
+                auto currentPosition = m_stream->tellg();
+
+                m_stream->seekg(0, std::ios_base::end);
+                int64_t retVal = -1;
+
+                if (m_stream->good())
                 {
-                    return NULL;
+                    retVal = static_cast<int64_t>(m_stream->tellg());
```

Comment:
again, I really don't know if you should be checking `good()` vs `operator bool()` because `good()` will be false once we've read to EOF (though I don't *think* seeking to end will set the eof flag?)

Reply:
it absolutely does, that's why we end up having to clear. but I'll definitely move it to the bool operator.